[ARMv7] Return exceptions from operations using long long

justinmichaud opened the pull request:
#### 3ebc10609bfec7c5b8e1da3236e141d2eb0ae13e
<pre>
[ARMv7] Return exceptions from operations using long long
<a href="https://bugs.webkit.org/show_bug.cgi?id=280196">https://bugs.webkit.org/show_bug.cgi?id=280196</a>

Reviewed by Keith Miller.

In JSToWasm, we now return exceptions via a register to make the ARMv7
port match the 64-bit ports. Returning an EncodedJSValue + an exception
is still not supported, so we modify the signatures of these JSToWasm
operations to return pointers or void instead.

We also fix a little issue with the templating logic, and add a static_assert
preventing it from happening again.

* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::operationExceptionRegister):
* Source/JavaScriptCore/jit/OperationResult.h:
(JSC::asExceptionResultBase):
(JSC::requires):
(JSC::ExceptionOperationImplicitResult::operator ExceptionOperationResultTag):
(JSC::ExceptionOperationImplicitResult::operator To):
(JSC::ExceptionOperationImplicitResult&lt;void&gt;::operator ExceptionOperationResultVoid):
(JSC::makeOperationResult):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
(JSC::Wasm::createJSToWasmJITShared):

Canonical link: <a href="https://commits.webkit.org/284329@main">https://commits.webkit.org/284329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a192a30ad3f90864ed7a9545a8a777d136f81246

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69050 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48450 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21722 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73131 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20208 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20057 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54975 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/13421 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44230 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59613 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35451 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40896 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17043 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18582 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62166 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17388 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74840 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68296 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13032 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16634 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62626 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13070 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59696 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62525 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15325 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10509 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4118 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90077 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44254 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15970 "Found 109 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Strings/property_and_index_of_string.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/sets.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck.js.default, ChakraCore.yaml/ChakraCore/test/strict/16.catch_sm.js.default, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-alias-one-block.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-alias.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-osr-exit-multiple-blocks.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-check-function-change-structure.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-convert-this-object-then-exit-on-other.js.layout ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45327 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46523 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45069 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jit/OperationResult.h:
```diff
@@ -138,6 +141,117 @@ ALWAYS_INLINE T makeOperationResult(Scope& scope, T result)
+
+template<typename T>
+static ALWAYS_INLINE ExceptionOperationResultTag asExceptionResult(T value, Exception* exception)
+requires (sizeof(T) == sizeof(uint32_t))
```
I was putting the `requires` clauses between the template and the signature. It seems like we should be consistent or change the other cases to match this.

kmiller68 commented on Source/JavaScriptCore/jit/OperationResult.h:
```diff
@@ -34,18 +34,21 @@
 namespace JSC {
 
+template<typename T>
+constexpr bool assert_not_operation_signature =
```
Seems like this shouldn't be snake case?

kmiller68 commented on Source/JavaScriptCore/jit/OperationResult.h:
```diff
@@ -34,18 +34,21 @@
 
+template<typename T>
+constexpr bool assert_not_operation_signature =
+    []{ static_assert(!std::is_function_v<typename std::remove_pointer_t<T>>, "This should take a return type, not an operator signature."); return true; }();
```
Why does this need to be wrapped in a lambda?

justinmichaud replied:
It needs to be an expression, static_assert is a statement. This makes sure that SFINAE can't just ignore it, which is what the previous code was doing.

justinmichaud replied:
Before, when you did operationExceptionRegister<operationBlah>, it would silently give you the non-void version

kmiller68 replied:
Can we change this to:
```
// This is wrapped in a lambda since it's used by static_asserts thus needs to be an expression.
template<typename T>
constexpr bool assertNotOperationSignature = ([] { 
    static_assert(!std::is_function_v<typename std::remove_pointer_t<T>>, "This should take a return type, not an operator signature."); 
    return true; 
})();
```

kmiller68 commented on Source/JavaScriptCore/jit/OperationResult.h:
```diff
@@ -138,6 +142,117 @@ ALWAYS_INLINE T makeOperationResult(Scope& scope, T result)
 
+#else // USE(JSVALUE64)
+
+enum class ExceptionOperationResultTag : uint64_t { };
```
Clever!

kmiller68 reviewed (approved):
r=me with one change

justinmichaud committed:
[ARMv7] Return exceptions from operations using long long
https://bugs.webkit.org/show_bug.cgi?id=280196

Reviewed by Keith Miller.

In JSToWasm, we now return exceptions via a register to make the ARMv7
port match the 64-bit ports. Returning an EncodedJSValue + an exception
is still not supported, so we modify the signatures of these JSToWasm
operations to return pointers or void instead.

We also fix a little issue with the templating logic, and add a static_assert
preventing it from happening again.

* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::operationExceptionRegister):
* Source/JavaScriptCore/jit/OperationResult.h:
(JSC::asExceptionResultBase):
(JSC::requires):
(JSC::ExceptionOperationImplicitResult::operator ExceptionOperationResultTag):
(JSC::ExceptionOperationImplicitResult::operator To):
(JSC::ExceptionOperationImplicitResult<void>::operator ExceptionOperationResultVoid):
(JSC::makeOperationResult):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
(JSC::Wasm::createJSToWasmJITShared):

Canonical link: https://commits.webkit.org/284329@main
